app/services/preserver.rb

```diff
@@ -0,0 +1,125 @@
+# frozen_string_literal: true
+class Preserver
+  # Provided as a mechanism for switching preservation strategies depending on
+  # the profile. For now, there's only one, so fall down to new.
+  def self.for(change_set:, change_set_persister:, storage_adapter: nil)
+    return NullPreserver unless change_set.try(:preserve?)
+    new(change_set: change_set, change_set_persister: change_set_persister, storage_adapter: storage_adapter)
+  end
+
+  class NullPreserver
+    def self.preserve!; end
+  end
+
+  attr_reader :change_set, :storage_adapter, :change_set_persister
+  delegate :resource, to: :change_set
+  def initialize(change_set:, storage_adapter: nil, change_set_persister:)
+    @change_set = change_set
+    @storage_adapter = storage_adapter || default_storage_adapter
+    @change_set_persister = change_set_persister
+  end
+
+  def preserve!
+    preserve_binary_content
+    # Don't preserve children unless this is the first time it's being
+    # preserved. After that point any updates to the children will trigger them
+    # to preserve themselves, because their parent is set up to be.
+    if preservation_object.persisted?
+      preserve_metadata
+    else
+      preserve_metadata && preserve_children
+    end
+  end
+
+  def preserve_binary_content
+    binary_nodes.each do |binary_node|
+      next if binary_node.file_identifiers.blank?
+      next if preservation_object.binary_nodes.find { |x| x.preservation_copy_of_id == binary_node.id }
+      file_metadata = file_metadata_from_binary_node(binary_node)
+      uploaded_file = storage_adapter.upload(
+        file: File.open(Valkyrie::StorageAdapter.find_by(id: binary_node.file_identifiers.first).disk_path),
+        original_filename: file_metadata.label.first,
+        resource: resource
+      )
+      file_metadata.file_identifiers = uploaded_file.id
+      preservation_object.binary_nodes ||= []
+      preservation_object.binary_nodes += [file_metadata]
+    end
+  end
+
+  def binary_nodes
+    [:original_files, :intermediate_files, :preservation_files].flat_map do |node_type|
+      Array(resource.try(node_type))
+    end
+  end
+
+  def file_metadata_from_binary_node(node)
+    label, splitter, extension = node.label.first.to_s.rpartition(".")
+    label = "#{label}-#{node.id}#{splitter}#{extension}"
```

Comment:
it's a bit hard to visualize what this will look like. Maybe you could give some comments with examples, either here or in the specs.